packages/core/src/subscriber/chainSubscribers.ts

```diff
@@ -0,0 +1,170 @@
+import {
+    services,
+    utils,
+    config,
+    contracts,
+    database,
+} from '../../';
+import { ethers } from 'ethers';
+
+class ChainSubscribers {
+    provider: ethers.providers.WebSocketProvider;
+    ifaceCommunityAdmin: ethers.utils.Interface;
+    filterTopics: string[][];
+
+    constructor(provider: ethers.providers.WebSocketProvider) {
+        this.provider = provider;
+        this.ifaceCommunityAdmin = new ethers.utils.Interface(
+            contracts.CommunityAdminABI
+        );
+
+        this.filterTopics = [
+            [
+                ethers.utils.id(
+                    'CommunityAdded(address,address[],uint256,uint256,uint256,uint256,uint256,uint256,uint256)'
+                ),
+                ethers.utils.id('CommunityRemoved(address)'),
+            ],
+        ];
+        this.recover();
+    }
+
+    stop() {
+        this.provider.removeAllListeners();
+        services.app.ImMetadataService.setRecoverBlockUsingLastBlock();
+    }
+
+    recover() {
+        this._setupListener(this.provider);
+        // we start the listener alongside with the recover system
+        // so we know we don't lose events.
+        this._runRecoveryTxs(this.provider).then(() =>
+            services.app.ImMetadataService.removeRecoverBlock()
+        )
+    }
+
+    async _runRecoveryTxs(
+        provider: ethers.providers.WebSocketProvider
+    ) {
+        utils.Logger.info('Recovering past events...');
+        let startFromBlock: number;
+        let lastBlockCached = await database.redisClient.get('lastBlock');
+        if (!lastBlockCached) {
+            startFromBlock = await services.app.ImMetadataService.getRecoverBlock();
+        } else {
+            startFromBlock = parseInt(lastBlockCached)
+        }
+
+        const rawLogs = await provider.getLogs({
+            fromBlock: startFromBlock,
+            toBlock: 'latest',
+            topics: this.filterTopics,
+        });
+
+        const logs = rawLogs.sort((a, b) => {
+            if (a.blockNumber > b.blockNumber) {
+                return 1;
+            }
+            if (a.blockNumber < b.blockNumber) {
+                return -1;
+            }
+            // a must be equal to b
+            return 0;
+        });
+
+        // iterate
+        for (let x = 0; x < logs.length; x += 1) {
+            // verify if cusd or community and do things
+            await this._filterAndProcessEvent(logs[x]);
+        }
+        utils.Logger.info('Past events recovered successfully!');
+    }
+
+    _setupListener(provider: ethers.providers.JsonRpcProvider) {
+        utils.Logger.info('Starting subscribers...');
+        const filter = {
+            topics: this.filterTopics,
+        };
+
+        database.redisClient.set('blockCount', 0);
+
+        provider.on(filter, async (log: ethers.providers.Log) => {
```

Comment:
Ahh, you might have mixed actually. This provider in `provider.on` is the on that should be a websocket.